crates/but-server/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+        app_settings: Arc::new(
+            AppSettingsWithDiskSync::new(config_dir.clone())
+                .expect("failed to create app settings"),
+        ),
+        user_controller: Arc::new(gitbutler_user::Controller::from_path(&app_data_dir)),
+        broadcaster: broadcaster.clone(),
+        archival: Arc::new(gitbutler_feedback::Archival {
+            cache_dir: app_data_dir.join("cache").clone(),
+            logs_dir: app_data_dir.join("logs").clone(),
+        }),
+    };
+
+    // build our application with a single route
+    let app = Router::new()
+        .route(
+            "/",
+            get(|| async { "Hello, World!" }).post({
+                let ipc_ctx = ipc_ctx.clone();
+                let extra = extra.clone();
+                move |req| handle_command(req, ipc_ctx, extra)
+            }),
+        )
+        .route(
+            "/ws",
+            any({
+                let broadcaster = broadcaster.clone();
+                async move |req| handle_ws_request(req, broadcaster).await
+            }),
+        )
+        .layer(ServiceBuilder::new().layer(cors));
+
+    // run our app with hyper, listening globally on port 6978
+    let port = std::env::var("BUTLER_PORT").unwrap_or("6978".into());
+    let host = std::env::var("BUTLER_HOST").unwrap_or("0.0.0.0".into());
+    let url = format!("{}:{}", host, port);
+    let listener = tokio::net::TcpListener::bind(&url).await.unwrap();
+    println!("Running at {}", url);
+    axum::serve(listener, app).await.unwrap();
+}
+
+async fn handle_ws_request(
+    ws: WebSocketUpgrade,
+    broadcaster: Arc<Mutex<Broadcaster>>,
+) -> impl IntoResponse {
+    ws.on_upgrade(move |socket| handle_websocket(socket, broadcaster))
+}
+
+async fn handle_websocket(socket: WebSocket, broadcaster: Arc<Mutex<Broadcaster>>) {
+    let (send, mut recv) = tokio::sync::mpsc::unbounded_channel();
+    let id = uuid::Uuid::new_v4();
+    broadcaster.lock().await.register_sender(&id, send);
+
+    let (mut socket_send, mut socket_recv) = socket.split();
+    let thread = tokio::spawn(async move {
+        while let Some(event) = recv.recv().await {
+            socket_send
+                .send(Message::Text(serde_json::to_string(&event).unwrap().into()))
+                .await
+                .unwrap();
+        }
+    });
+
+    while let Some(Ok(msg)) = socket_recv.next().await {
+        #[allow(clippy::single_match)]
+        match msg {
+            Message::Close(_) => {
+                thread.abort();
+                break;
+            }
+            _ => {}
+        }
+    }
+
+    broadcaster.lock().await.deregister_sender(&id);
+}
+
+fn run_cmd<
+    D: DeserializeOwned,
+    S: Serialize,
+    Fun: Fn(&IpcContext, D) -> Result<S, but_api::error::Error>,
+>(
+    ipc_ctx: &IpcContext,
+    params: serde_json::Value,
+    fun: Fun,
+) -> Result<serde_json::Value, but_api::error::Error> {
+    let result = fun(ipc_ctx, serde_json::from_value(params).to_error()?)?;
```

Comment:
Here it might be interesting to know what happens if `fun` panics. Does this bring down the server? If so, maybe that's what we want initially.

Reply:
The thread panics, but the server remains running - which I think is the desired behaviour